src/Storages/MergeTree/ReplicatedMergeTreeRestartingThread.cpp

```diff
@@ -58,106 +58,108 @@ void ReplicatedMergeTreeRestartingThread::run()
-
-                if (!need_stop && !tryStartup())
-                {
-                    /// We couldn't startup replication. Table must be readonly.
-                    /// Otherwise it can have partially initialized queue and other
-                    /// strange parts of state.
-                    setReadonly();
-
-                    if (first_time)
-                        storage.startup_event.set();
-
-                    task->scheduleAfter(retry_period_ms);
-                    return;
-                }
-
-                if (first_time)
-                    storage.startup_event.set();
-
-                startup_completed = true;
-            }
-
-            if (need_stop)
-                return;
-
-            bool old_val = true;
-            if (storage.is_readonly.compare_exchange_strong(old_val, false))
-            {
-                readonly_mode_was_set = false;
-                CurrentMetrics::sub(CurrentMetrics::ReadonlyReplica);
-            }
-
-            first_time = false;
-        }
+        bool replica_is_active = runImpl();
+        if (!replica_is_active)
+            reschedule_period_ms = retry_period_ms;
     }
     catch (const Exception & e)
     {
         /// We couldn't activate table let's set it into readonly mode
-        setReadonly();
         partialShutdown();
-        storage.startup_event.set();
         tryLogCurrentException(log, __PRETTY_FUNCTION__);
 
         if (e.code() == ErrorCodes::REPLICA_STATUS_CHANGED)
-            reschedule_now = true;
+            reschedule_period_ms = 0;
     }
     catch (...)
     {
-        setReadonly();
         partialShutdown();
-        storage.startup_event.set();
         tryLogCurrentException(log, __PRETTY_FUNCTION__);
     }
 
-    if (reschedule_now)
+    if (first_time)
+    {
+        if (storage.is_readonly)
+        {
+            /// We failed to start replication, table is still readonly, so we should increment the metric. See also setNotReadonly().
+            CurrentMetrics::add(CurrentMetrics::ReadonlyReplica);
+        }
+        /// It does not matter if replication is actually started or not, just notify after the first attempt.
+        storage.startup_event.set();
+        first_time = false;
+    }
+
+    if (need_stop)
+        return;
+
+    if (reschedule_period_ms)
+        task->scheduleAfter(reschedule_period_ms);
+    else
         task->schedule();
+}
+
+bool ReplicatedMergeTreeRestartingThread::runImpl()
+
+{
+    if (!storage.is_readonly && !storage.getZooKeeper()->expired())
+        return true;
+
+    if (first_time)
+    {
+        LOG_DEBUG(log, "Activating replica.");
+        assert(storage.is_readonly);
+    }
+    else if (storage.is_readonly)
+    {
+        LOG_WARNING(log, "Table was in readonly mode. Will try to activate it.");
+    }
+    else if (storage.getZooKeeper()->expired())
+    {
+        LOG_WARNING(log, "ZooKeeper session has expired. Switching to a new session.");
+        partialShutdown();
```

Comment:
Question: Is `partialShutdown` idempotent? Can we call it multiple times?

Reply:
Yep